lib/ansible/modules/cloud/amazon/ec2_instance.py

```
@@ -0,0 +1,1300 @@
+                resp = ec2.terminate_instances(InstanceIds=[id_])
+                [changed.add(i['InstanceId']) for i in resp['TerminatingInstances']]
+            if desired_state == 'STOPPED':
+                resp = ec2.stop_instances(InstanceIds=[id_])
+                [changed.add(i['InstanceId']) for i in resp['StoppingInstances']]
+            if desired_state == 'RUNNING':
+                resp = ec2.start_instances(InstanceIds=[id_])
+                [changed.add(i['InstanceId']) for i in resp['StartingInstances']]
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError):
+            # we don't care about exceptions here, as we'll fail out if any instances failed to terminate
+            pass
+
+    if changed:
+        await_instances(ids=list(changed), state=desired_state)
+
+    change_failed = list(to_change - changed)
+    instances = find_instances(ec2, ids=list(to_change))
+    return changed, change_failed, instances
+
+
+def pretty_instance(i):
+    instance = camel_dict_to_snake_dict(i, ignore_list=['Tags'])
+    instance['tags'] = boto3_tag_list_to_ansible_dict(i['Tags'])
+    return instance
+
+
+def determine_iam_role(name_or_arn, iam):
+    if re.match(r'^arn:aws:iam::\d+:instance-profile/[\w+=/,.@-]+$', name_or_arn):
+        return name_or_arn
+    if iam is None:
+        iam = module.client('iam')
+    try:
+        role = iam.get_instance_profile(InstanceProfileName=name_or_arn)
+        return role['InstanceProfile']['Arn']
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchEntity':
+            module.fail_json_aws(msg="Could not find instance_role {0}".format(name_or_arn))
+        module.fail_json_aws(msg="An error occurred while searching for instance_role "
+                                 "{0}. Please try supplying the full ARN.".format(name_or_arn))
+    module.fail_json(msg="Could not retrieve instance_role {0}".format(name_or_arn))
```

I think if `return role['InstanceProfile']['Arn']` fails, that should raise an Exception, right? I don't think this line gets used.

---

Correct. Removed.